torch/csrc/jit/codegen/cuda/index_compute.cpp

```diff
@@ -1939,6 +1887,241 @@ std::pair<std::vector<kir::Val*>, bool> Index::getConsumerRootPredIndices(
+
+  std::unordered_map<kir::IterDomain*, kir::Val*> reference_halo_extent_map;
+
+  const auto& halo_info = gpu_lower->haloInfo();
+
+  // Generate map from reference iter domains to halo extents
+  for (auto entry : ref_2_consumer) {
+    auto ref_id = entry.first;
+    auto extent = halo_info.getExtent(ref_id);
+    if (extent != nullptr) {
+      reference_halo_extent_map[gpu_lower->lowerValue(ref_id)
+                                    ->as<kir::IterDomain>()] =
+          gpu_lower->lowerValue(extent);
+    }
+  }
+
+  // Index into the reference tensor
+  auto ref_indexing = getReferenceIndexing(
+      loops,
+      reference_domain,
+      ref_id_to_ind_map,
+      {},
+      reference_halo_extent_map);
+
+  // If we are initializing a reduction buffer and the tensor has a
+  // rfactor root, the predicate should be based on the rfactor root.
+  const auto root_domain = reference_domain->getRootDomain();
+
+  auto contig_id_infos = getPredicateContigIds(reference_domain->domain());
+
+  // Roots in contiguous processing is based on reference roots, want to convert
+  // these to concrete roots, flip reference's concrete_to_id map
+  decltype(reference_id_map) ref_id_to_concrete;
+  std::transform(
+      reference_id_map.begin(),
+      reference_id_map.end(),
+      std::inserter(ref_id_to_concrete, ref_id_to_concrete.begin()),
+      [](auto entry) { return std::make_pair(entry.second, entry.first); });
+
+  std::vector<std::unordered_set<IterDomain*>> handeled_roots;
+
+  std::vector<kir::Bool*> predicates;
+
+  for (auto contig_id_entry : contig_id_infos) {
+    auto contig_id = contig_id_entry.contig_id;
+    if (contig_id->isBroadcast() ||
+        gpu_lower->trivialReductionInfo().isDerived(contig_id)) {
+      continue;
+    }
+
+    auto root_ids = contig_id_entry.root_ids;
+    auto kir_contig_id =
+        gpu_lower->lowerValue(contig_id)->as<kir::IterDomain>();
+
+    const auto it = ref_indexing.indexMap().find(kir_contig_id);
+    if (it == ref_indexing.indexMap().end() || it->second->isZeroInt()) {
+      continue;
+    }
+
+    auto extent = kir_contig_id->extent();
+
+    auto halo_extent_it = reference_halo_extent_map.find(kir_contig_id);
+    if (halo_extent_it != reference_halo_extent_map.end()) {
+      extent = halo_extent_it->second;
+    }
+
+    if (it == ref_indexing.indexMap().end() || it->second->isZeroInt() ||
+        (it->second->definition() == nullptr &&
+         extent->definition() == nullptr)) {
+      continue;
+    }
+
+    predicates.push_back(
+        ir_builder.ltExpr(it->second, extent)->as<kir::Bool>());
+
+    std::unordered_set<IterDomain*> concrete_root_ids;
+    std::transform(
+        contig_id_entry.root_ids.begin(),
+        contig_id_entry.root_ids.end(),
+        std::inserter(concrete_root_ids, concrete_root_ids.begin()),
+        [&ref_id_to_concrete](IterDomain* root_id) {
+          return ref_id_to_concrete.at(root_id);
+        });
```

Comment:
`PredicateContigInfo::root_ids` seems to be concrete root IDs, if I understand what's done in the loop at line 1936. Is this transform really necessary?

Reply:
Yes, my mistake. Must've implemented this on different days 😳

Reply:
The transformation on 1936 is actually unnecessary. When I wrote that part I forgot that reference IDs are not part of the compute at mappings. You have to explicitly use the mapping with the reference tensor. So this transformation is needed, the other is not.